### ISSUE TITLE
GroupKFold not working as it should

### ISSUE THREAD

**Issue by jeroenvermunt** [pycaret/pycaret#2259]
**Describe the bug**
I am using pycaret to get an overview of the accuracy scores that I can achieve with my current dataset. For this task to goal is to predict whether a case will be won or lose, making it a binary classification problem. Each case has multiple observations with the same win/loss label. It is crucial that multiple observations of each case are not in both the train and test set. To ensure that this happens, I used groupKFold as custom cross-validation. However, I am achieving unrealistically high scores on my prediction and am suspecting this is caused by the cross validation not working as it should.

```python
s = setup(data, target = "won", fold_strategy='groupkfold', fold_groups='index')
best = compare_models(n_select=5)
plot_model(best[0],'feature')
plot_model(best[3],'feature')
```

**Interpretation**
The code results in: (I blurred out some feature names for privacy)
![image](https://user-images.githubusercontent.com/39826623/157430629-ea877523-b55a-47a2-ab3c-702d290a9fbc.png)
![image](https://user-images.githubusercontent.com/39826623/157430829-5e6ba673-4cf2-487c-8b03-3e639ad0106a.png)
![image](https://user-images.githubusercontent.com/39826623/157431492-0d5609fd-e0d1-484e-b1a6-73208ff65b82.png)

The accuracy scores are unrealistically high, suggesting that something has gone wrong. Looking at the feature importances of the best model we see the index feature, which is what I specified as the group for GroupKFold. This suggests that something has gone wrong with the GroupKFold splitter. The fourth best model does not show the index feature, but also suffers from a similarly high score.

**Expected behavior**
I expect the cross-validation to occur using GroupKFold from sklearn. However, looking at the result it seems like that has not occured even though I specified it in the setup. I expect the accuracies to be around 60-80% and that index is not present in the feature importances. Is this a bug or did I do something wrong in the setup?

**Versions**
pycaret version 2.3.6

**Comment by moezali1** [comment]
@Yard1 Can you take a look at this one when you get a chance? I doubt it's a bug but want to be sure.

**Comment by Yard1** [comment]
Will do

**Comment by JelleAdan** [comment]
I have exactly the same problem in pycaret! 

Trained a model and got >0.99 accuracy on classifier models when using "groupkfold" as the fold strategy. Then decided to use scikitlearn random forest classifier and also got a >0.99 accuracy after cross validation. However, when using the option "groupkfold" as fold strategy here, the accuracy dropped to approximately 0.75.

Seems like this option fails in pycaret...

**Comment by sorenwacker** [comment]
I have something similar with pycaret 2.3.10. It seems pycaret fails to remove the feature used as fold_groups from the features list and uses it as a categorical feature. The following setup had no categorical features, but one is used:

![image](https://user-images.githubusercontent.com/3391614/164309192-f0181788-31d5-424d-9cc2-57f0b9f625c7.png)

**Comment by sorenwacker** [comment]
Compare this to the case when providing the groups directly:
![image](https://user-images.githubusercontent.com/3391614/164309550-4799d13e-f110-412d-ab57-6babeae0f43a.png)

However, this also does not work:

![image](https://user-images.githubusercontent.com/3391614/164309706-07c98aa8-5a97-4317-8fb8-ffef9883f3c2.png)

Notice that using `GroupKFold` forces the user to use `imputation_type='iterative'`.

Maybe I am using it wrong? But no model is created.

**Comment by sorenwacker** [comment]
Apparently, the problem here was the use of `data_split_stratify=True`. Setting it to `False` finally generated models.

**Comment by ekamperi** [comment]
Hey folks. Any update on this one? This is a significant blocker for us, and I'm surprised it wasn't prioritized. Can we help somehow to get this fixed? Like offer a bounty or help with testing? Thanks!

**Comment by Yard1** [comment]
Can you try with the 3.0 prerelease version and see if the problem still occurs? Thanks.

**Comment by ekamperi** [comment]
With 3.0 prerelease `compare_models()` returns an empty list here, when setting `fold_strategy='groupkfold'` and `fold_groups` set either to the column name or to an array of samples.

**Comment by Yard1** [comment]
Can you provide a reproductible code sample?

**Comment by Jorgvt** [comment]
I also found that the column used in `fold_groups` was still being used, but used `ignore_features` to ignore it during training and seems that it's doing it properly.

**Comment by moezali1** [comment]
@tvdboom Is this resolved now as well?

**Comment by tvdboom** [comment]
This is resolved.

> EDIT: Also found out that if you don't set data_split_shuffle=False, the data will be shuffled before separating in groups and thus the indexes provided by the GroupedKFold won't be correct.

In rc5, you don't need to set `data_split_shuffle=False`, but you have to specify `index=True` to leave the rows with their original index after shuffling
